sympy/solvers/solveset.py

```diff
@@ -3478,19 +3485,45 @@ def _separate_poly_nonpoly(system, symbols):
             polys_expr.append(poly.as_expr())
         else:
             nonpolys.append(eq)
-    return polys, polys_expr, nonpolys, denominators
+    return polys, polys_expr, nonpolys, denominators, unrad_changed
 # end of def _separate_poly_nonpoly()
 
 
+def _handle_poly(polys, symbols):
+    # Compute a Groebner basis in lex order wrt the ordering given by
+    # symbols.
+    #
+    # XXX: Maybe it is possible to choose a better ordering of the symbols
+    # here like choose the symbol was highest (or lowest?) degree first...
+
+    basis = groebner(polys, symbols, polys=False)
```

Comment:
It seems that that `fglm` cannot convert a positive-dimensional Groebner basis to lex, and passing a Groebner basis in grevlex order to `substitution` (which is done in the positive-dimensional case) is causing it to return spurious solutions in some cases. Currently, the fix I have implemented is to recompute the Groebner basis in lex order in the positive-dimensional case. Do you think it is worth it to first compute the basis in grevlex order and then convert it to lex in the zero-dimensional case considering the additional cost of recomputing the basis in lex order for the positive dimensional case?

Reply:
I tried comparing the time taken by `groebner` to compute a basis for a positive-dimensional system with the time taken by `substitution` to solve the system (with the basis passed instead of the original system) with some random systems to see if the extra time needed for recomputing the basis in the positive dimensional case was significant as compared to the time taken by `substitution` to actually solve the system. In most cases, `substitution` was significantly slower than `groebner` (usually by more than an order of magnitude) and many times did not complete within a few minutes. Also, in some cases `substitution` terminated with errors that look like bugs. 
I think a condition can be added to check if the grevlex Groebner basis contains just `1` to avoid recomputing the basis in the case where there are no solutions. Otherwise, I think it shouldn't make too much of a performance difference to recompute the basis. Also, on master, the Groebner basis is computed twice anyways in the positive-dimensional case, once to check if the system is zero dimensional and again in `_handle_positive_dimensional`, so the current implementation can only offer a speedup even in the positive-dimensional case. @oscarbenjamin What do you think?

Reply:
I think that sounds reasonable.

Another possibility that occurs to me is that sometimes the Groebner basis can be factored which can lead to multiple simpler Grobner bases for substitution to solve. Here is an example:
```python
In [64]: polys = [y + x, x**2 + y*z**2 + t, (z + t)*(z + 1)]

In [65]: gb = groebner(polys)

In [66]: gb
Out[66]: 
             ⎛⎡                           2                   2    ⎤                                 ⎞
GroebnerBasis⎝⎣x + y, -t⋅y⋅z - t⋅y + t + y  - y⋅z, t⋅z + t + z  + z⎦, x, y, z, t, domain=ℤ, order=lex⎠

In [67]: gb[-1].factor()
Out[67]: (t + z)⋅(z + 1)

In [68]: f1, f2 = gb[-1].factor().args

In [69]: groebner(gb[:-1] + [f1])
Out[69]: 
             ⎛⎡            2           ⎤                                 ⎞
GroebnerBasis⎝⎣x + y, t + y  + y, z + 1⎦, x, y, z, t, domain=ℤ, order=lex⎠

In [70]: groebner(gb[:-1] + [f2])
Out[70]: 
             ⎛⎡        2          2       ⎤                                 ⎞
GroebnerBasis⎝⎣x + y, t ⋅y + t + y , t + z⎦, x, y, z, t, domain=ℤ, order=lex⎠

```

Reply:
> I think that sounds reasonable.

Ok. The latest commit implements the approach I suggested.

> Another possibility that occurs to me is that sometimes the Groebner basis can be factored which can lead to multiple simpler Grobner bases for substitution to solve. Here is an example:

I tried implementing this strategy, including splitting the original basis into multiple bases if multiple terms can be factored, but it seems that it is not as straightforward as I originally thought. For one thing, if multiple polynomials in the basis can be factored, it can quickly lead to a large number of simpler Groebner bases if all possible combinations of one factor from each polynomial are considered. In this situation, it is not clear to me how many and which polynomials should be used for factoring to maximize the performance gain while retaining correctness.
Also, in some cases, the solution from one simpler basis can be a special case of the solution from another one, which would have to be handled correctly. For example,
```python
>>> sys = [a -d, b + d, c + d, (d - 1)*(c*d-1)]
>>> gb = groebner(sys)
>>> gb
GroebnerBasis([a - d, b + d, c + d, d**3 - d**2 + d - 1], a, b, c, d, domain='ZZ', order='lex')
>>> f1, f2 = gb[-1].factor().args
>>> gb1 = groebner(gb[:-1] + [f1])
>>> gb2 = groebner(gb[:-1] + [f2])
>>> substitution(gb1, [a, b, c, d])
{(d, -d, -d, d)}
>>> substitution(gb2, [a, b, c, d])
{(1, -1, -1, 1)}
```
Here, the second solution is a special case of the first and shouldn't probably be included in the final solution. 

Unless an implementation of this strategy already exists somewhere in the codebase, I don't think this will be particularly easy to implement. I think in this PR we can maybe just go with the current approach.

Reply:
I just realized that the first solution (`(d, -d, -d, d)`) is incorrect in the example above. That is the reason for the redundancy in the solutions. I think the redundant solution issue could still occur in some cases (for e.g. solving `[a + b, (c-d)*(c-1)]` for `[a, b, c]`) even if `substitution` does not produce any wrong results. However, considering that `substitution` seems to frequently give incorrect solutions, it is possible this strategy could currently lead to a lot of spurious solutions, as each call to `substitution` increases the chances of incorrect solutions creeping in even with relatively simple systems. I guess `substitution` needs to be fixed before this approach can be implemented.

Reply:
There are a couple of bugs leading to the wrong result for `gb1`:
```python
In [1]: a, b, c, d = symbols('a:d')

In [2]: eqs = [a - d, b + d, c + d, d**2 + 1]

In [3]: substitution(eqs, [a, b, c, d])
Out[3]: {(d, -d, -d, d)}
```


Firstly `solve_using_known_values` ignores an EmptySet when it tries to solve `d**2 + 1` for real solutions and then it generates `(d, -d, -d, d)` as a solution along with the two complex solutions so the set of solutions looks like this:
```python
(Pdb) p result_all_variables
[{b: -d, c: -d, a: d, d: d}, {d: I, b: -I, c: -I, a: I}, {d: -I, b: I, c: I, a: -I}]
```
Then here the two (correct) complex solutions are dropped:
https://github.com/sympy/sympy/blob/ff5755cecd165cfde49adbe6b4b976dc486fccf3/sympy/solvers/solveset.py#L3394-L3398
It should not be presumed that an infinite solution component is necessarily a superset of all finite solution components. It's not hard to see how that can lead to throwing away part of the solution:
```python
In [6]: nonlinsolve([a*(a - log(b)), a*(b-2)], [a, b])
Out[6]: {(0, b)}
```
That leaves two bugs to be fixed:

1. If `solve_using_known_values` finds EmptySet for any equation it should return EmptySet for the whole system.
2. Finite solutions should not be thrown away because there is also an infinite solution.

This diff addresses both bugs but does result in some redundant solutions. It's better to have redundant solutions than remove incorrect solutions though:
```diff
diff --git a/sympy/solvers/solveset.py b/sympy/solvers/solveset.py
index ff934ae..74164f6 100644
--- a/sympy/solvers/solveset.py
+++ b/sympy/solvers/solveset.py
@@ -3259,6 +3259,11 @@ def _solve_using_known_values(result, solver):
                     try:
                         soln = solver(eq2, sym)
                         total_solvest_call += 1
+
+                        if soln is S.EmptySet:
+                            breakpoint()
+                            return [], total_solvest_call, total_conditionst
+
                         soln_new = S.EmptySet
                         if isinstance(soln, Complement):
                             # separate solution and complement
@@ -3274,11 +3279,6 @@ def _solve_using_known_values(result, solver):
                                 intersections[sym] = soln.args[0]
                             soln_new += soln.args[1]
                         soln = soln_new if soln_new else soln
-                        if index > 0 and solver == solveset_real:
-                            # one symbol's real soln, another symbol may have
-                            # corresponding complex soln.
-                            if not isinstance(soln, (ImageSet, ConditionSet)):
-                                soln += solveset_complex(eq2, sym)  # might give ValueError with Abs
                     except (NotImplementedError, ValueError):
                         # If solveset is not able to solve equation `eq2`. Next
                         # time we may get soln using next equation `eq2`
@@ -3391,11 +3391,6 @@ def _solve_using_known_values(result, solver):
         if res not in result_all_variables:
             result_all_variables.append(res)
 
-    if result_infinite:
-        # we have general soln
-        # eg : [{x: -1, y : 1}, {x : -y, y: y}] then
-        # return [{x : -y, y : y}]
-        result_all_variables = result_infinite
     if intersections or complements:
         result_all_variables = add_intersection_complement(
             result_all_variables, intersections, complements)
```
Redundancy (`(0, 2)` is redundant but `(log(2), 2)` is not):
```python
In [1]: a, b, c, d = symbols('a:d')

In [2]: eqs = [a - d, b + d, c + d, d**2 + 1]

In [3]: nonlinsolve([a*(a - log(b)), a*(b-2)], [a, b])
Out[3]: {(0, 2), (0, b), (log(2), 2)}
```
Also with the diff:
```python
In [1]: a, b, c, d = symbols('a:d')

In [2]: eqs = [a - d, b + d, c + d, d**2 + 1]

In [3]: nonlinsolve(eqs, [a, b, c, d])
Out[3]: {(-ⅈ, ⅈ, ⅈ, -ⅈ), (ⅈ, -ⅈ, -ⅈ, ⅈ)}
```

Reply:
The diff you suggest causes some test failures (mostly due to redundant or incomplete solutions), which will probably have to be addressed. I think a new PR should be created for implementing these changes. I'll try to get to that once I am done working adding the tests for the systems mentioned in this PR as you suggest.